File: python-app/vault_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Vault Python Client
hvac 라이브러리를 사용한 Vault 클라이언트 구현
"""

import time
import logging
from typing import Dict, Any, Optional, Tuple
import hvac
from hvac.exceptions import VaultError


class VaultClient:
    """Vault 클라이언트 클래스"""
    
    def __init__(self, config: Dict[str, Any]):
        """
        Vault 클라이언트 초기화
        
        Args:
            config: Vault 설정 정보
        """
        self.config = config
        self.client = None
        self.token = None
        self.token_issued_time = 0
        self.token_ttl = 0
        
        # 캐시 저장소
        self.kv_cache = {}
        self.db_dynamic_cache = {}
        self.db_static_cache = {}
        
        # 로깅 설정
        self.logger = logging.getLogger(__name__)
        
        # Vault 클라이언트 초기화
        self._init_client()
    
    def _init_client(self):
        """Vault 클라이언트 초기화"""
        try:
            self.client = hvac.Client(
                url=self.config['url'],
                namespace=self.config.get('namespace')
            )
            self.logger.info(f"Vault 클라이언트 초기화 완료: {self.config['url']}")
        except Exception as e:
            self.logger.error(f"Vault 클라이언트 초기화 실패: {e}")
            raise
    
    def login(self) -> bool:
        """
        AppRole을 사용한 Vault 로그인
        
        Returns:
            로그인 성공 여부
        """
        try:
            response = self.client.auth.approle.login(
                role_id=self.config['role_id'],
                secret_id=self.config['secret_id']
            )
            
            self.token = response['auth']['client_token']
            self.token_issued_time = time.time()
            self.token_ttl = response['auth']['lease_duration']
            
            # 클라이언트에 토큰 설정
            self.client.token = self.token
            
            self.logger.info(f"Vault 로그인 성공 (TTL: {self.token_ttl}초)")
            return True
            
        except VaultError as e:
            self.logger.error(f"Vault 로그인 실패: {e}")
            return False
        except Exception as e:
            self.logger.error(f"로그인 중 오류 발생: {e}")
            return False
    
    def renew_token(self) -> bool:
        """
        토큰 갱신
        
        Returns:
            갱신 성공 여부
        """
        try:
            response = self.client.auth.token.renew_self()
            
            self.token_issued_time = time.time()
            self.token_ttl = response['auth']['lease_duration']
            
            self.logger.info(f"토큰 갱신 성공 (TTL: {self.token_ttl}초)")
            return True
            
        except VaultError as e:
            self.logger.error(f"토큰 갱신 실패: {e}")
            return False
        except Exception as e:
            self.logger.error(f"토큰 갱신 중 오류 발생: {e}")
            return False
    
    def is_token_expired(self) -> bool:
        """
        토큰 만료 여부 확인
        
        Returns:
            토큰 만료 여부
        """
        if not self.token or self.token_ttl <= 0:
            return True
        
        elapsed_time = time.time() - self.token_issued_time
        # TTL의 4/5 지점에서 갱신
        return elapsed_time >= (self.token_ttl * 0.8)
    
    def ensure_valid_token(self) -> bool:
        """
        유효한 토큰 보장
        
        Returns:
            토큰 유효성 보장 성공 여부
        """
        if not self.token or self.is_token_expired():
            return self.login()
        
        if self.is_token_expired():
            return self.renew_token()
        
        return True
    
    def get_kv_secret(self, path: str) -> Optional[Dict[str, Any]]:
        """
        KV v2 시크릿 조회
        
        Args:
            path: 시크릿 경로
            
        Returns:
            시크릿 데이터 또는 None
        """
        if not self.ensure_valid_token():
            return None
        
        try:
            # 캐시 확인
            if path in self.kv_cache:
                cached_data = self.kv_cache[path]
                current_time = time.time()
                
                # 캐시가 유효한지 확인 (5분)
                if current_time - cached_data['timestamp'] < 300:
                    self.logger.debug(f"KV 시크릿 캐시 사용: {path}")
                    return cached_data['data']
            
            # Vault에서 시크릿 조회
            response = self.client.secrets.kv.v2.read_secret_version(
                path=path,
                mount_point=f"{self.config['entity']}-kv"
            )
            
            secret_data = response['data']['data']
            metadata = response['data']['metadata']
            
            # 캐시 저장
            self.kv_cache[path] = {
                'data': secret_data,
                'metadata': metadata,
                'timestamp': time.time()
            }
            
            self.logger.info(f"KV 시크릿 조회 성공 (버전: {metadata['version']})")
            return secret_data
            
        except VaultError as e:
            self.logger.error(f"KV 시크릿 조회 실패: {e}")
            return None
        except Exception as e:
            self.logger.error(f"KV 시크릿 조회 중 오류 발생: {e}")
            return None
    
    def get_database_dynamic_secret(self, role_id: str) -> Optional[Dict[str, Any]]:
        """
        Database Dynamic 시크릿 조회
        
        Args:
            role_id: Database 역할 ID
            
        Returns:
            시크릿 데이터 또는 None
        """
        if not self.ensure_valid_token():
            return None
        
        try:
            # 캐시 확인
            cache_key = f"{role_id}"
            if cache_key in self.db_dynamic_cache:
                cached_data = self.db_dynamic_cache[cache_key]
                current_time = time.time()
                
                # TTL 기반 캐시 확인 (10초 임계값)
                remaining_ttl = cached_data['ttl'] - (current_time - cached_data['timestamp'])
                if remaining_ttl > 10:
                    self.logger.debug(f"Database Dynamic 시크릿 캐시 사용 (TTL: {int(remaining_ttl)}초)")
                    return {
                        'data': cached_data['data'],
                        'ttl': int(remaining_ttl)
                    }
            
            # Vault에서 시크릿 조회
            response = self.client.secrets.database.generate_credentials(
                name=role_id,
                mount_point=f"{self.config['entity']}-database"
            )
            
            secret_data = response['data']
            ttl = response['lease_duration']
            
            # 캐시 저장
            self.db_dynamic_cache[cache_key] = {
                'data': secret_data,
                'ttl': ttl,
                'timestamp': time.time()
            }
            
            self.logger.info(f"Database Dynamic 시크릿 조회 성공 (TTL: {ttl}초)")
            return {
                'data': secret_data,
                'ttl': ttl
            }
            
        except VaultError as e:
            self.logger.error(f"Database Dynamic 시크릿 조회 실패: {e}")
            return None
        except Exception as e:
            self.logger.error(f"Database Dynamic 시크릿 조회 중 오류 발생: {e}")
            return None
    
    def get_database_static_secret(self, role_id: str) -> Optional[Dict[str, Any]]:
        """
        Database Static 시크릿 조회
        
        Args:
            role_id: Database 역할 ID
            
        Returns:
            시크릿 데이터 또는 None
        """
        if not self.ensure_valid_token():
            return None
        
        try:
            # 캐시 확인
            cache_key = f"{role_id}"
            if cache_key in self.db_static_cache:
                cached_data = self.db_static_cache[cache_key]
                current_time = time.time()
                
                # 시간 기반 캐시 확인 (5분)
                if current_time - cached_data['timestamp'] < 300:
                    self.logger.debug(f"Database Static 시크릿 캐시 사용")
                    return cached_data['data']
            
            # Vault에서 시크릿 조회
            response = self.client.secrets.database.get_static_credentials(
                name=role_id,
                mount_point=f"{self.config['entity']}-database"
            )
            
            secret_data = response['data']
            ttl = response.get('ttl', 3600)  # 기본 TTL 1시간
            
            # 캐시 저장
            self.db_static_cache[cache_key] = {
                'data': secret_data,
                'ttl': ttl,
                'timestamp': time.time()
            }
            
            self.logger.info(f"Database Static 시크릿 조회 성공 (TTL: {ttl}초)")
            return {
                'data': secret_data,
                'ttl': ttl
            }
            
        except VaultError as e:
            self.logger.error(f"Database Static 시크릿 조회 실패: {e}")
            return None
        except Exception as e:
            self.logger.error(f"Database Static 시크릿 조회 중 오류 발생: {e}")
            return None
    
    def get_token_info(self) -> Optional[Dict[str, Any]]:
        """
        토큰 정보 조회
        
        Returns:
            토큰 정보 또는 None
        """
        if not self.ensure_valid_token():
            return None
        
        try:
            response = self.client.auth.token.lookup_self()
            return response['data']
        except VaultError as e:
            self.logger.error(f"토큰 정보 조회 실패: {e}")
            return None
        except Exception as e:
            self.logger.error(f"토큰 정보 조회 중 오류 발생: {e}")
            return None


if __name__ == "__main__":
    """Vault 클라이언트 테스트"""
    import sys
    from config_loader import VaultConfig
    
    # 로깅 설정
    logging.basicConfig(level=logging.INFO)
    
    try:
        # 설정 로드
        config_loader = VaultConfig()
        vault_config = config_loader.get_vault_config()
        
        # Vault 클라이언트 생성
        client = VaultClient(vault_config)
        
        # 로그인 테스트
        if client.login():
            print("✅ Vault 로그인 성공")
            
            # 토큰 정보 조회
            token_info = client.get_token_info()
            if token_info:
                print(f"토큰 정보: {token_info}")
        else:
            print("❌ Vault 로그인 실패")
            sys.exit(1)
            
    except Exception as e:
        print(f"테스트 실패: {e}")
        sys.exit(1)

File: samples/c-sharp/VaultClient.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using VaultSharp;
using VaultSharp.V1.AuthMethods.AppRole;
using VaultSharp.V1.Commons;

namespace VaultCsharpApp
{
    /// <summary>
    /// Vault 클라이언트 클래스
    /// VaultSharp 라이브러리 사용
    /// </summary>
    public class VaultClient
    {
        private readonly VaultSettings _config;
        private readonly ILogger<VaultClient> _logger;

        private IVaultClient? _client;
        private string? _token;
        private long   _tokenIssuedTime;  // Unix timestamp (초)
        private int    _tokenTtl;

        // ── 캐시 ────────────────────────────────────────────────
        private readonly Dictionary<string, KvCacheEntry>        _kvCache        = new();
        private readonly Dictionary<string, DbDynamicCacheEntry> _dbDynamicCache = new();
        private readonly Dictionary<string, DbStaticCacheEntry>  _dbStaticCache  = new();

        /// <summary>
        /// Vault 클라이언트 초기화
        /// </summary>
        /// <param name="config">Vault 서버 연결 설정</param>
        /// <param name="logger">로거 인스턴스</param>
        public VaultClient(VaultSettings config, ILogger<VaultClient> logger)
        {
            _config = config;
            _logger = logger;
            InitClient();
        }

        // ── 초기화 ───────────────────────────────────────────────

        private void InitClient()
        {
            _logger.LogInformation("Vault 클라이언트 초기화 완료: {Url} (Namespace: {Namespace})",
                _config.Url, _config.Namespace ?? "(없음)");
        }

        // ── 인증 ────────────────────────────────────────────────

        /// <summary>
        /// AppRole을 사용한 Vault 로그인
        /// </summary>
        /// <returns>로그인 성공 여부</returns>
        public async Task<bool> LoginAsync()
        {
            try
            {
                var authMethod = new AppRoleAuthMethodInfo(_config.RoleId, _config.SecretId);
                var vaultClientSettings = new VaultClientSettings(_config.Url, authMethod)
                {
                    MyHttpClientProviderFunc = null,
                    // Namespace가 null이면 네임스페이스 미사용 (Community Edition 호환)
                    Namespace = _config.Namespace
                };

                _client = new VaultSharp.VaultClient(vaultClientSettings);

                // 실제 로그인 확인을 위해 토큰 조회
                var tokenInfo = await _client.V1.Auth.Token.LookupSelfAsync();
                _token = tokenInfo.Data.Id;
                _tokenIssuedTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
                _tokenTtl = 3600; // 기본 TTL 1시간

                _logger.LogInformation("Vault 로그인 성공 (TTL: {Ttl}초)", _tokenTtl);
                return true;
            }
            catch (Exception e)
            {
                _logger.LogError("Vault 로그인 실패: {Error}", e.Message);
                return false;
            }
        }

        /// <summary>
        /// 토큰 갱신
        /// </summary>
        /// <returns>갱신 성공 여부</returns>
        public async Task<bool> RenewTokenAsync()
        {
            if (_client == null) return false;
            try
            {
                await _client.V1.Auth.Token.RenewSelfAsync();
                _tokenIssuedTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
                _logger.LogInformation("토큰 갱신 성공");
                return true;
            }
            catch (Exception e)
            {
                _logger.LogError("토큰 갱신 실패: {Error}", e.Message);
                return false;
            }
        }

        /// <summary>
        /// 토큰 만료 여부 확인 (TTL의 80% 지점에서 갱신)
        /// </summary>
        /// <returns>토큰 만료 여부</returns>
        public bool IsTokenExpired()
        {
            if (_token == null || _tokenTtl <= 0) return true;
            var elapsed = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - _tokenIssuedTime;
            // TTL의 80% 지점에서 갱신
            return elapsed >= _tokenTtl * 0.8;
        }

        /// <summary>
        /// 유효한 토큰 보장
        /// </summary>
        /// <returns>토큰 유효성 보장 성공 여부</returns>
        public async Task<bool> EnsureValidTokenAsync()
        {
            if (_token == null || IsTokenExpired())
                return await LoginAsync();
            return true;
        }

        // ── 시크릿 조회 ─────────────────────────────────────────

        /// <summary>
        /// KV v2 시크릿 조회 (5분 캐시)
        /// </summary>
        /// <param name="path">시크릿 경로 (마운트 포인트 제외)</param>
        /// <returns>시크릿 데이터 또는 null</returns>
        public async Task<Dictionary<string, object>?> GetKvSecretAsync(string path)
        {
            if (!await EnsureValidTokenAsync()) return null;

            // 캐시 확인 (5분 유효)
            if (_kvCache.TryGetValue(path, out var cached))
            {
                var age = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - cached.Timestamp;
                if (age < 300)
                {
                    _logger.LogDebug("KV 시크릿 캐시 사용: {Path}", path);
                    return cached.Data;
                }
            }

            try
            {
                // 마운트 포인트: {entity}-kv (예: my-vault-app-kv)
                var mountPoint = $"{_config.Entity}-kv";
                var secret = await _client!.V1.Secrets.KeyValue.V2
                    .ReadSecretAsync(path: path, mountPoint: mountPoint);

                var data = secret.Data.Data?.ToDictionary(k => k.Key, v => v.Value)
                           ?? new Dictionary<string, object>();

                // 캐시 저장
                _kvCache[path] = new KvCacheEntry
                {
                    Data      = data,
                    Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
                };

                _logger.LogInformation("KV 시크릿 조회 성공 (버전: {Ver})", secret.Data.Metadata.Version);
                return data;
            }
            catch (Exception e)
            {
                _logger.LogError("KV 시크릿 조회 실패: {Error}", e.Message);
                return null;
            }
        }

        /// <summary>
        /// Database Dynamic 시크릿 조회 (TTL 기반 캐시, 10초 임계값)
        /// </summary>
        /// <param name="roleId">Vault에 등록된 Database 역할 이름</param>
        /// <returns>시크릿 데이터(username/password)와 TTL, 또는 null</returns>
        public async Task<SecretResult?> GetDatabaseDynamicSecretAsync(string roleId)
        {
            if (!await EnsureValidTokenAsync()) return null;

            // 캐시 확인 (TTL 기반, 10초 임계값)
            if (_dbDynamicCache.TryGetValue(roleId, out var cached))
            {
                var remainingTtl = cached.Ttl - (DateTimeOffset.UtcNow.ToUnixTimeSeconds() - cached.Timestamp);
                if (remainingTtl > 10)
                {
                    _logger.LogDebug("Database Dynamic 시크릿 캐시 사용 (TTL: {Ttl}초)", remainingTtl);
                    return new SecretResult { Data = cached.Data, Ttl = (int)remainingTtl };
                }
            }

            try
            {
                // 마운트 포인트: {entity}-database (예: my-vault-app-database)
                var mountPoint = $"{_config.Entity}-database";
                var secret = await _client!.V1.Secrets.Database
                    .GetCredentialsAsync(roleId, mountPoint: mountPoint);

                var data = new Dictionary<string, string>
                {
                    ["username"] = secret.Data.Username,
                    ["password"] = secret.Data.Password
                };

                var ttl = (int)secret.LeaseDurationSeconds;

                // 캐시 저장
                _dbDynamicCache[roleId] = new DbDynamicCacheEntry
                {
                    Data      = data,
                    Ttl       = ttl,
                    Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
                };

                _logger.LogInformation("Database Dynamic 시크릿 조회 성공 (TTL: {Ttl}초)", ttl);
                return new SecretResult { Data = data, Ttl = ttl };
            }
            catch (Exception e)
            {
                _logger.LogError("Database Dynamic 시크릿 조회 실패: {Error}", e.Message);
                return null;
            }
        }

        /// <summary>
        /// Database Static 시크릿 조회 (5분 캐시)
        /// </summary>
        /// <param name="roleId">Vault에 등록된 Database 역할 이름</param>
        /// <returns>시크릿 데이터(username/password)와 TTL, 또는 null</returns>
        public async Task<SecretResult?> GetDatabaseStaticSecretAsync(string roleId)
        {
            if (!await EnsureValidTokenAsync()) return null;

            // 캐시 확인 (5분 유효)
            if (_dbStaticCache.TryGetValue(roleId, out var cached))
            {
                var age = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - cached.Timestamp;
                if (age < 300)
                {
                    _logger.LogDebug("Database Static 시크릿 캐시 사용");
                    return cached.Result;
                }
            }

            try
            {
                // 마운트 포인트: {entity}-database (예: my-vault-app-database)
                var mountPoint = $"{_config.Entity}-database";
                var secret = await _client!.V1.Secrets.Database
                    .GetStaticCredentialsAsync(roleId, mountPoint: mountPoint);

                var data = new Dictionary<string, string>
                {
                    ["username"] = secret.Data.Username,
                    ["password"] = secret.Data.Password
                };

                var ttl = (int)(secret.LeaseDurationSeconds > 0 ? secret.LeaseDurationSeconds : 3600);
                var result = new SecretResult { Data = data, Ttl = ttl };

                // 캐시 저장
                _dbStaticCache[roleId] = new DbStaticCacheEntry
                {
                    Result    = result,
                    Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
                };

                _logger.LogInformation("Database Static 시크릿 조회 성공 (TTL: {Ttl}초)", ttl);
                return result;
            }
            catch (Exception e)
            {
                _logger.LogError("Database Static 시크릿 조회 실패: {Error}", e.Message);
                return null;
            }
        }
    }

    // ── 결과/캐시 모델 ──────────────────────────────────────────

    /// <summary>시크릿 조회 결과</summary>
    public class SecretResult
    {
        /// <summary>시크릿 데이터 (username, password 등)</summary>
        public Dictionary<string, string> Data { get; set; } = new();
        /// <summary>남은 TTL (초)</summary>
        public int Ttl { get; set; }
    }

    /// <summary>KV 시크릿 캐시 항목</summary>
    internal class KvCacheEntry
    {
        public Dictionary<string, object> Data      { get; set; } = new();
        public long                        Timestamp { get; set; }
    }

    /// <summary>Database Dynamic 시크릿 캐시 항목</summary>
    internal class DbDynamicCacheEntry
    {
        public Dictionary<string, string> Data      { get; set; } = new();
        public int                         Ttl       { get; set; }
        public long                        Timestamp { get; set; }
    }

    /// <summary>Database Static 시크릿 캐시 항목</summary>
    internal class DbStaticCacheEntry
    {
        public SecretResult Result    { get; set; } = new();
        public long         Timestamp { get; set; }
    }
}
